### ISSUE TITLE
Ability to issue diagnostics during projection

### ISSUE THREAD

**Issue by tjprescott** [microsoft/typespec#2492]
**TL/DR:** The ask is for a mechanism to emit diagnostics during projection.

I am working on a fix for #1244.

I can do this validation very easily inside of `projection-members.ts`. The logic is simple--you examine the new name and make sure there isn't already a property in the model that has that name. However, all I can do in this situation is throw a `ProjectionError`, which causes a huge ugly wall of text and breaks all compilation (well, more accurately it breaks the emit process).

![image](https://github.com/microsoft/typespec/assets/5723682/9e33ac87-d16e-4c8e-8c35-a1173b6b0973)

The alternative is to add logic to `$onValidate`, which runs once during compilation. The problem is we have to build a ton of convoluted, hard to understand and maintain code to figure out what the post-projection _should_ look like in order to generate an error diagnostic (duplicate-property). If the actual projection logic changes but this "projection emulation" code doesn't, shenanigans ensue.

This is pretty much the standard procedure for all things Versioning and contributes to it being buggy and hard to maintain.
